src/storage/mutate/DeleteVertexProcessor.cpp

```diff
@@ -17,54 +17,131 @@ void DeleteVertexProcessor::process(const cpp2::DeleteVertexRequest& req) {
-        return;
+        VLOG(3) << "Error! ret = " << static_cast<int32_t>(ret)
+                << ", spaceId " << spaceId;
+        return "";
     }
+    std::vector<std::string> indexes;
     while (iter->valid()) {
         auto key = iter->key();
+        auto tagId = NebulaKeyUtils::getTagId(key);
         if (FLAGS_enable_vertex_cache && vertexCache_ != nullptr) {
             if (NebulaKeyUtils::isVertex(key)) {
-                auto tagId = NebulaKeyUtils::getTagId(key);
                 VLOG(3) << "Evict vertex cache for vId " << vId << ", tagId " << tagId;
                 vertexCache_->evict(std::make_pair(vId, tagId), partId);
             }
         }
-        keys.emplace_back(key.data(), key.size());
+        batchHolder->remove(key.str());
+        for (auto& index : indexes_) {
+            auto indexId = index.get_index_id();
+            if (index.get_tagOrEdge() == tagId) {
+                auto reader = RowReader::getTagPropReader(this->schemaMan_,
+                                                          iter->val(),
+                                                          spaceId,
+                                                          tagId);
+                const auto& cols = index.get_cols();
+                auto values = collectIndexValues(reader.get(), cols);
+                auto indexKey = NebulaKeyUtils::vertexIndexKey(partId,
+                                                               indexId,
+                                                               vId,
+                                                               values);
+                std::string val;
+                auto result = kvstore_->get(spaceId, partId, indexKey, &val);
+                if (result == kvstore::ResultCode::SUCCEEDED) {
+                    indexes.emplace_back(std::move(indexKey));
+                }
+            }
+        }
         iter->next();
     }
-
-    this->kvstore_->asyncMultiRemove(spaceId,
-                                     partId,
-                                     std::move(keys),
-                                     [spaceId, partId, this](kvstore::ResultCode code) {
-        VLOG(3) << "partId:" << partId << ", code:" << static_cast<int32_t>(code);
-        cpp2::ResultCode thriftResult;
-        thriftResult.set_code(to(code));
-        thriftResult.set_part_id(partId);
-        if (code == kvstore::ResultCode::ERR_LEADER_CHANGED) {
-            nebula::cpp2::HostAddr leader;
-            auto addrRet = kvstore_->partLeader(spaceId, partId);
-            CHECK(ok(addrRet));
-            auto addr = value(std::move(addrRet));
-            leader.set_ip(addr.first);
-            leader.set_port(addr.second);
-            thriftResult.set_leader(std::move(leader));
+    /**
+     * There may be different versions of the same vertex,
+     * so delete duplicate index keys is required.
+     **/
+    if (!indexes.empty()) {
+        sort(indexes.begin(), indexes.end());
```

Comment:
We should skip the duplicate versions key when iterating.

Reply:
Make sense!